### PR TITLE
fix nll loss decomposition to properly ignore ignore_index

### LINKED ISSUE #95794: CUDA IMA on forward pass of BertForMaskedLM with PT2 RC
### 🐛 Describe the bug

When running the forward pass of a transformers `BertForMaskedLM` with `torch.compile`, an illegal memory CUDA error appears. The same code runs fine without `torch.compile`-ing the model.

Repro:
```py
import torch

import transformers
from transformers import AutoModelForMaskedLM, AutoTokenizer, DataCollatorForLanguageModeling

def main():
    model_name = "bert-base-uncased"
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForMaskedLM.from_pretrained(model_name)
    
    texts = ["This is a text for the example."] * 16
    tokenized_texts = tokenizer(texts, padding="max_length", truncation=True, max_length=128, return_tensors="pt")      
    data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer)
    batch = tokenized_texts
    batch["input_ids"], batch["labels"] = data_collator.torch_mask_tokens(batch["input_ids"])

    model = torch.compile(model, backend="inductor")
    model.to("cuda")

    batch = {k: v.to("cuda") for k, v in batch.items()}
    outputs = model(**batch)
    loss = outputs.loss
    loss.backward()


if __name__ == "__main__":
    main()
```

The script works fine when compiling with `"aot_eager"`.

### Error logs

```
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /home/sylvain_huggingface_co/git/transformers/test_mlm.py:27 in <module>                         │
│                                                                                                  │
│   24                                                                                             │
│   25                                                                                             │
│   26 if __name__ == "__main__":                                                                  │
│ ❱ 27 │   main()                                                                                  │
│   28                                                                                             │
│                                                                                                  │
│ /home/sylvain_huggingface_co/git/transformers/test_mlm.py:23 in main                             │
│                                                                                                  │
│   20 │   batch = {k: v.to("cuda") for k, v in batch.items()}                                     │
│   21 │   outputs = model(**batch)                                                                │
│   22 │   loss = outputs.loss                                                                     │
│ ❱ 23 │   loss.backward()                                                                         │
│   24                                                                                             │
│   25                                                                                             │
│   26 if __name__ == "__main__":                                                                  │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_tensor.py:487 in backward │
│                                                                                                  │
│    484 │   │   │   │   create_graph=create_graph,                                                │
│    485 │   │   │   │   inputs=inputs,                                                            │
│    486 │   │   │   )                                                                             │
│ ❱  487 │   │   torch.autograd.backward(                                                          │
│    488 │   │   │   self, gradient, retain_graph, create_graph, inputs=inputs                     │
│    489 │   │   )                                                                                 │
│    490                                                                                           │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/autograd/__init__.py:204   │
│ in backward                                                                                      │
│                                                                                                  │
│   201 │   # The reason we repeat same the comment below is that                                  │
│   202 │   # some Python versions print out the first line of a multi-line function               │
│   203 │   # calls in the traceback and some print out the last line                              │
│ ❱ 204 │   Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the bac   │
│   205 │   │   tensors, grad_tensors_, retain_graph, create_graph, inputs,                        │
│   206 │   │   allow_unreachable=True, accumulate_grad=True)  # Calls into the C++ engine to ru   │
│   207                                                                                            │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/autograd/function.py:274   │
│ in apply                                                                                         │
│                                                                                                  │
│   271 │   │   │   │   │   │   │      "Function is not allowed. You should only implement one "   │
│   272 │   │   │   │   │   │   │      "of them.")                                                 │
│   273 │   │   user_fn = vjp_fn if vjp_fn is not Function.vjp else backward_fn                    │
│ ❱ 274 │   │   return user_fn(self, *args)                                                        │
│   275 │                                                                                          │
│   276 │   def apply_jvp(self, *args):                                                            │
│   277 │   │   # _forward_cls is defined by derived class                                         │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py │
│ :2358 in backward                                                                                │
│                                                                                                  │
│   2355 │   │   │   │   # Pass args even though they're unused, so that the graph is built        │
│   2356 │   │   │   │   out = CompiledFunctionBackward.apply(*all_args)                           │
│   2357 │   │   │   else:                                                                         │
│ ❱ 2358 │   │   │   │   out = call_compiled_backward()                                            │
│   2359 │   │   │   return out                                                                    │
│   2360 │                                                                                         │
│   2361 │   compiled_function = create_runtime_wrapper(                                           │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py │
│ :2329 in call_compiled_backward                                                                  │
│                                                                                                  │
│   2326 │   │   │   │   │   else:                                                                 │
│   2327 │   │   │   │   │   │   context = disable_autocast_manager if disable_amp else nullconte  │
│   2328 │   │   │   │   │   │   with context(), track_graph_compiling(aot_config, "backward"):    │
│ ❱ 2329 │   │   │   │   │   │   │   CompiledFunction.compiled_bw = aot_config.bw_compiler(        │
│   2330 │   │   │   │   │   │   │   │   bw_module, all_args                                       │
│   2331 │   │   │   │   │   │   │   )                                                             │
│   2332                                                                                           │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_dynamo/backends/common.py │
│ :38 in _wrapped_bw_compiler                                                                      │
│                                                                                                  │
│    35 │   │                                                                                      │
│    36 │   │   def _wrapped_bw_compiler(*args, **kwargs):                                         │
│    37 │   │   │   # stop TorchDynamo from trying to compile our generated backwards pass         │
│ ❱  38 │   │   │   return eval_frame.disable(eval_frame.disable(bw_compiler)(*args, **kwargs))    │
│    39 │   │                                                                                      │
│    40 │   │   bw_compiler = kwargs.get("bw_compiler") or kwargs["fw_compiler"]                   │
│    41 │   │   kwargs["bw_compiler"] = _wrapped_bw_compiler                                       │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py:228  │
│ in _fn                                                                                           │
│                                                                                                  │
│   225 │   │   │   dynamic_ctx = enable_dynamic(self.dynamic)                                     │
│   226 │   │   │   dynamic_ctx.__enter__()                                                        │
│   227 │   │   │   try:                                                                           │
│ ❱ 228 │   │   │   │   return fn(*args, **kwargs)                                                 │
│   229 │   │   │   finally:                                                                       │
│   230 │   │   │   │   set_eval_frame(prior)                                                      │
│   231 │   │   │   │   dynamic_ctx.__exit__(None, None, None)                                     │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_dynamo/utils.py:164 in    │
│ time_wrapper                                                                                     │
│                                                                                                  │
│    161 │   │   │   if key not in compilation_metrics:                                            │
│    162 │   │   │   │   compilation_metrics[key] = []                                             │
│    163 │   │   │   t0 = time.time()                                                              │
│ ❱  164 │   │   │   r = func(*args, **kwargs)                                                     │
│    165 │   │   │   time_spent = time.time() - t0                                                 │
│    166 │   │   │   # print(f"Dynamo timer: key={key}, latency={latency:.2f} sec")                │
│    167 │   │   │   compilation_metrics[key].append(time_spent)                                   │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/compile_fx.py:47 │
│ 3 in bw_compiler                                                                                 │
│                                                                                                  │
│   470 │   @dynamo_utils.dynamo_timed                                                             │
│   471 │   def bw_compiler(model: torch.fx.GraphModule, example_inputs):                          │
│   472 │   │   fixed = count_tangents(model)                                                      │
│ ❱ 473 │   │   return inner_compile(                                                              │
│   474 │   │   │   model,                                                                         │
│   475 │   │   │   example_inputs,                                                                │
│   476 │   │   │   num_fixed=fixed,                                                               │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_dynamo/debug_utils.py:595 │
│ in debug_wrapper                                                                                 │
│                                                                                                  │
│    592 │   │   │   compiled_fn = deferred_for_real_inputs                                        │
│    593 │   │   │   compiled_fn._boxed_call = True                                                │
│    594 │   │   else:                                                                             │
│ ❱  595 │   │   │   compiled_fn = compiler_fn(gm, example_inputs)                                 │
│    596 │   │                                                                                     │
│    597 │   │   return compiled_fn                                                                │
│    598                                                                                           │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/debug.py:239 in  │
│ inner                                                                                            │
│                                                                                                  │
│   236 │   │   @functools.wraps(fn)                                                               │
│   237 │   │   def inner(*args, **kwargs):                                                        │
│   238 │   │   │   with DebugContext():                                                           │
│ ❱ 239 │   │   │   │   return fn(*args, **kwargs)                                                 │
│   240 │   │                                                                                      │
│   241 │   │   return wrap_compiler_debug(inner, compiler_name="inductor")                        │
│   242                                                                                            │
│                                                                                                  │
│ /usr/lib/python3.8/contextlib.py:75 in inner                                                     │
│                                                                                                  │
│    72 │   │   @wraps(func)                                                                       │
│    73 │   │   def inner(*args, **kwds):                                                          │
│    74 │   │   │   with self._recreate_cm():                                                      │
│ ❱  75 │   │   │   │   return func(*args, **kwds)                                                 │
│    76 │   │   return inner                                                                       │
│    77                                                                                            │
│    78                                                                                            │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/compile_fx.py:18 │
│ 0 in compile_fx_inner                                                                            │
│                                                                                                  │
│   177 │   │   )                                                                                  │
│   178 │   │   with V.set_graph_handler(graph):                                                   │
│   179 │   │   │   graph.run(*example_inputs)                                                     │
│ ❱ 180 │   │   │   compiled_fn = graph.compile_to_fn()                                            │
│   181 │                                                                                          │
│   182 │   if cudagraphs:                                                                         │
│   183 │   │   complex_memory_overlap_inputs = any(                                               │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/graph.py:609 in  │
│ compile_to_fn                                                                                    │
│                                                                                                  │
│   606 │   │   return mod                                                                         │
│   607 │                                                                                          │
│   608 │   def compile_to_fn(self):                                                               │
│ ❱ 609 │   │   return self.compile_to_module().call                                               │
│   610 │                                                                                          │
│   611 │   def get_output_names(self):                                                            │
│   612 │   │   assert self.graph_outputs is not None                                              │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_dynamo/utils.py:164 in    │
│ time_wrapper                                                                                     │
│                                                                                                  │
│    161 │   │   │   if key not in compilation_metrics:                                            │
│    162 │   │   │   │   compilation_metrics[key] = []                                             │
│    163 │   │   │   t0 = time.time()                                                              │
│ ❱  164 │   │   │   r = func(*args, **kwargs)                                                     │
│    165 │   │   │   time_spent = time.time() - t0                                                 │
│    166 │   │   │   # print(f"Dynamo timer: key={key}, latency={latency:.2f} sec")                │
│    167 │   │   │   compilation_metrics[key].append(time_spent)                                   │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/graph.py:598 in  │
│ compile_to_module                                                                                │
│                                                                                                  │
│   595 │   │   if config.debug:                                                                   │
│   596 │   │   │   print(code)                                                                    │
│   597 │   │                                                                                      │
│ ❱ 598 │   │   mod = PyCodeCache.load(code)                                                       │
│   599 │   │   for name, value in self.constants.items():                                         │
│   600 │   │   │   setattr(mod, name, value)                                                      │
│   601                                                                                            │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/codecache.py:603 │
│ in load                                                                                          │
│                                                                                                  │
│   600 │   │   │   │   mod = types.ModuleType(f"{__name__}.{key}")                                │
│   601 │   │   │   │   mod.__file__ = path                                                        │
│   602 │   │   │   │   mod.key = key                                                              │
│ ❱ 603 │   │   │   │   exec(code, mod.__dict__, mod.__dict__)                                     │
│   604 │   │   │   │   # another thread might set this first                                      │
│   605 │   │   │   │   cls.cache.setdefault(key, mod)                                             │
│   606 │   │   return cls.cache[key]                                                              │
│                                                                                                  │
│ /tmp/torchinductor_sylvain_huggingface_co/52/c52y7w7bmhzx2y2qe3snv7oy2xtpr7mahn5f4jvhtaapzzjkc7a │
│ u.py:1005 in <module>                                                                            │
│                                                                                                  │
│   1002 ''')                                                                                      │
│   1003                                                                                           │
│   1004                                                                                           │
│ ❱ 1005 async_compile.wait(globals())                                                             │
│   1006 del async_compile                                                                         │
│   1007                                                                                           │
│   1008 def call(args):                                                                           │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/codecache.py:790 │
│ in wait                                                                                          │
│                                                                                                  │
│   787 │   │   │   │   if config.verbose_progress and not isinstance(pbar, _Faketqdm):            │
│   788 │   │   │   │   │   pbar.set_postfix_str(key)                                              │
│   789 │   │   │   │   if isinstance(result, (Future, TritonFuture)):                             │
│ ❱ 790 │   │   │   │   │   scope[key] = result.result()                                           │
│   791 │   │   │   │   │   pbar.update(1)                                                         │
│   792 │   │                                                                                      │
│   793 │   │   _compile_end()                                                                     │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/codecache.py:649 │
│ in result                                                                                        │
│                                                                                                  │
│   646 │   │   │   return self.kernel                                                             │
│   647 │   │   # If the worker failed this will throw an exception.                               │
│   648 │   │   self.future.result()                                                               │
│ ❱ 649 │   │   kernel = self.kernel = _load_kernel(self.source_code)                              │
│   650 │   │   latency = time() - t0                                                              │
│   651 │   │   if latency > 50:                                                                   │
│   652 │   │   │   name = _load_kernel_name(self.source_code)                                     │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/codecache.py:629 │
│ in _load_kernel                                                                                  │
│                                                                                                  │
│   626                                                                                            │
│   627 def _load_kernel(source_code):                                                             │
│   628 │   kernel = TritonCodeCache.load(source_code)                                             │
│ ❱ 629 │   kernel.precompile()                                                                    │
│   630 │   return kernel                                                                          │
│   631                                                                                            │
│   632                                                                                            │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/triton_ops/autot │
│ une.py:67 in precompile                                                                          │
│                                                                                                  │
│    64 │   │   with self.lock:                                                                    │
│    65 │   │   │   if self.launchers:                                                             │
│    66 │   │   │   │   return                                                                     │
│ ❱  67 │   │   │   self.launchers = [                                                             │
│    68 │   │   │   │   self._precompile_config(c, warm_cache_only_with_cc)                        │
│    69 │   │   │   │   for c in self.configs                                                      │
│    70 │   │   │   ]                                                                              │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/triton_ops/autot │
│ une.py:68 in <listcomp>                                                                          │
│                                                                                                  │
│    65 │   │   │   if self.launchers:                                                             │
│    66 │   │   │   │   return                                                                     │
│    67 │   │   │   self.launchers = [                                                             │
│ ❱  68 │   │   │   │   self._precompile_config(c, warm_cache_only_with_cc)                        │
│    69 │   │   │   │   for c in self.configs                                                      │
│    70 │   │   │   ]                                                                              │
│    71 │   │   │   self.configs = None                                                            │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/_inductor/triton_ops/autot │
│ une.py:92 in _precompile_config                                                                  │
│                                                                                                  │
│    89 │   │   # load binary to the correct device                                                │
│    90 │   │   with torch.cuda.device(compile_meta["device"]):                                    │
│    91 │   │   │   # need to initialize context                                                   │
│ ❱  92 │   │   │   torch.cuda.synchronize(torch.cuda.current_device())                            │
│    93 │   │   │   binary = triton.compile(                                                       │
│    94 │   │   │   │   self.fn,                                                                   │
│    95 │   │   │   │   **compile_meta,                                                            │
│                                                                                                  │
│ /home/sylvain_huggingface_co/dynamo/lib/python3.8/site-packages/torch/cuda/__init__.py:711 in    │
│ synchronize                                                                                      │
│                                                                                                  │
│    708 │   """                                                                                   │
│    709 │   _lazy_init()                                                                          │
│    710 │   with torch.cuda.device(device):                                                       │
│ ❱  711 │   │   return torch._C._cuda_synchronize()                                               │
│    712                                                                                           │
│    713                                                                                           │
│    714 def ipc_collect():                                                                        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

### Minified repro

_No response_

### Versions

Collecting environment information...
PyTorch version: 2.0.0.dev20230301+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.8.10 (default, Nov 14 2022, 12:59:47)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-1027-gcp-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 515.86.01
cuDNN version: Probably one of the following:
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          12
On-line CPU(s) list:             0-11
Thread(s) per core:              2
Core(s) per socket:              6
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                        7
CPU MHz:                         2200.216
BogoMIPS:                        4400.43
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       192 KiB
L1i cache:                       192 KiB
L2 cache:                        6 MiB
L3 cache:                        38.5 MiB
NUMA node0 CPU(s):               0-11
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; Clear CPU buffers; SMT Host state unknown
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.0.0+b8b470bc59
[pip3] torch==2.0.0.dev20230301+cu117
[pip3] torchaudio==0.13.1
[pip3] torchtriton==2.0.0+0d7e753227
[pip3] torchvision==0.15.0.dev20221122+cpu
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #95794
This is a hotfix for decomposition only (that is currently used by inductor), reference still accesses invalid indices. Perhaps `_nll_loss_nd` and this decomp should be unified, cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire @lezcano 

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -2845,35 +2844,31 @@ def nll_loss_forward(
...
+        total_weight = (target != ignore_index).sum().to(self)
 
     if reduction == Reduction.SUM.value:
         result = result.sum()
     elif reduction == Reduction.MEAN.value:
         if weight is None:
-            result = result.sum() / total_weight if ignore_index >= 0 else result.mean()
+            result = result.sum() / total_weight
         else:
             result = result.sum() / total_weight
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:2873]: You can now get rid of the branches.

**Review by lezcano — APPROVED** [review]
Ugh, this is annoying.
As per the docs, the previous implementation was right and correct. But I guess that some models were doing something like using a value outside of the range [0, C-1] and then ignored it via `ignore_index`. This is not specified in the docs, as it says that `target[i] \in [0, C-1]`.

It would be better then to have `ignore_index = None` by default, and then use this info to perform the optimisation that later was done via `ignore_index >= 0`. But of course, this would be divergent from core, which sets `ignore_index = -100` (???). Oh my.

Also, this is a great example of a decomposition whose derivative could be generated automatically via the Python dispatcher.

Anyway, the changes LGTM.

**Review by Chillee — APPROVED** [review]

**Comment by ngimel** [comment]
Docs also say
>if ignore_index is specified, this loss also accepts this class index (this index may not necessarily be in the class range).

**Comment by lezcano** [comment]
ugh fine. Still, it's annoying we don't have `None` as a default, as in these cases we could avoid the data-dependent paths

**Comment by ngimel** [comment]
Backward for `index`/`gather` is `index_put(..accum=True)`/`scatter_add` whereas here we use our knowledge that there's a single index to be gathered to use `scatter` instead in backward. `scatter_add` would generate a code with atomic_add, which probably would be fine, no collisions, but why bother?

**Comment by lezcano** [comment]
Even more, nothing is data dependent. Scratch everything I've said above. I'm going to bed.

**Pull request closed without merging**